include/deal.II/multigrid/mg_transfer_interface.h

```diff
@@ -0,0 +1,424 @@
+
+private:
+  /**
+   * Partitioner needed by an intermediate vector, which is needed for
+   * collecting all degrees of freedom of the children cells.
+   */
+  std::shared_ptr<const Utilities::MPI::Partitioner> extended_partitioner;
+
+  /**
+   * Indices for copying the data from/to the intermediate vector.
+   */
+  std::vector<unsigned int> indices;
+
+  friend class MGTransferUtil::Implementation;
+};
+
+
+
+#ifndef DOXYGEN
+
+/* ----------------------- Inline functions --------------------------------- */
+
+
+
+template <typename Number>
+template <typename Stream>
+void
+TransferScheme<Number>::print(Stream &out) const
+{
+  out << "weights:" << std::endl;
+  for (const auto w : weights)
+    out << w << " ";
+  out << std::endl;
+
+  out << "level_dof_indices_fine:" << std::endl;
+  for (const auto w : level_dof_indices_fine)
+    out << w << " ";
+  out << std::endl;
+
+  out << "level_dof_indices_coarse:" << std::endl;
+  for (const auto w : level_dof_indices_coarse)
+    out << w << " ";
+  out << std::endl;
+
+  out << "prolongation_matrix_1d:" << std::endl;
+  for (const auto w : prolongation_matrix_1d)
+    out << w[0] << " ";
+  out << std::endl;
+}
+
+
+
+template <int dim, typename Number>
+template <typename Stream>
+void
+Transfer<dim, Number>::print_internal(Stream &out) const
+{
+  for (const auto &scheme : schemes)
+    scheme.print(out);
+}
+
+
+
+template <typename MatrixType>
+MGTransferMatrixFreeNew<MatrixType>::MGTransferMatrixFreeNew(
+  const MGLevelObject<MatrixType> &           matrices,
+  const MGLevelObject<Transfer<dim, Number>> &transfer)
+  : matrices(matrices)
+  , transfer(transfer)
+{}
+
+
+
+template <typename MatrixType>
+void
+MGTransferMatrixFreeNew<MatrixType>::prolongate(
+  const unsigned int                                        to_level,
+  dealii::LinearAlgebra::distributed::Vector<Number> &      dst,
+  const dealii::LinearAlgebra::distributed::Vector<Number> &src) const
+{
+  this->transfer[to_level].prolongate(0 /*dummy*/, dst, src);
+}
+
+
+
+template <typename MatrixType>
+void
+MGTransferMatrixFreeNew<MatrixType>::restrict_and_add(
+  const unsigned int                                        from_level,
+  dealii::LinearAlgebra::distributed::Vector<Number> &      dst,
+  const dealii::LinearAlgebra::distributed::Vector<Number> &src) const
+{
+  this->transfer[from_level].restrict_and_add(0 /*dummy*/, dst, src);
+}
+
+
+
+template <typename MatrixType>
+template <class InVector, int spacedim>
+void
+MGTransferMatrixFreeNew<MatrixType>::copy_to_mg(
+  const DoFHandler<dim, spacedim> &dof_handler,
+  MGLevelObject<dealii::LinearAlgebra::distributed::Vector<Number>> &dst,
+  const InVector &                                                   src) const
+{
+  (void)dof_handler;
+
+  for (unsigned int level = dst.min_level(); level <= dst.max_level(); ++level)
+    matrices[level].initialize_dof_vector(dst[level]);
+
+  dst[dst.max_level()].copy_locally_owned_data_from(src);
```

Comment:
why do you only need to touch the finest vector?

Reply:
Why should you fill the vectors on the levels? They are filled during restriction, aren't they?